### ISSUE TITLE
warn about missing requires and imports

### ISSUE THREAD

**Issue by borkdude** [borkdude/clj-kondo#339]
Warn about the following:

```
(ns foo)
(clojure.string/starts-with? ...)
```
;;=> missing require (or import) for clojure.string

```
(ns bar)
(foo/x ...)
```
;;=> unknown alias foo

although we cannot know if foo is supposed to be an alias, a single-segmented namespace, or Java classname.

We would have to exclude known things like `goog` in CLJS.
Clj-kondo already have a list of all the by default available classes like `System`.

The thing that should be very well researched is false positives.
There's already a function for detecting Java class names:
 https://github.com/borkdude/clj-kondo/blob/2b4229ea652021e0e5ec6c73082fcc4c7e35d5f2/src/clj_kondo/impl/namespace.clj#L152

TODO:

- [x] exclude js/ and goog/ in CLJS
- [x] lint foo/bar usage, not only in call position
- [x] think more about project.clj defproject kind of macros. can fully qualified symbols occur in macros without proper requires / imports? in fact, user/defproject is the only macro that I need to include for now. Maybe special case it.

Note: the current workaround doesn't work in emacs, since the filename is not included
- [x] change linter name to something more neutral like 'unresolved-namespace' which is more factual. The message could be something like 'Unresolved namespaced. Did you forget a require?`.

Maybe later:

- [x] in comment sections, handle requires as if they are top-level

**Comment by borkdude** [comment]
TODO:
- [ ] rename `:unqualified?` to `:unresolved?` in result of `resolve-name`

**Comment by borkdude** [comment]
borkdude [5:17 PM]
 Hmm. In Clojure when you write:
```(ns foo)
bar/baz`
```
`bar/baz` could in theory be a public field in a class `bar` right? so it's hard to say this is an unresolved symbol by linting statically
bronsa [5:19 PM]
 there's no ambiguity between `Klass/field` and `ns/var`
borkdude [5:19 PM]
 so yes?
in CLJS-land there are also things like `goog/isFunction`
that are just "there" without requiring
bronsa [5:21 PM]
 yes, w/o keeping a runtime mapping of the available classes it's not easy to distinguish the two
borkdude [5:21 PM]
 currently clj-kondo doesn't report anything with a namespace as unresolved, but thinking about this more, I think it's the correct behavior because of the above

(given that underreporting is better than overreporting)

**Comment by borkdude** [comment]
Related #515

**Comment by markus-wa** [comment]
Hi. I was wondering what the rationale is for expecting a `require` for all used namespaces.

I'm fairly new to Clojure and I couldn't find this being suggested elsewhere. The code seems to work with or without require.

PS: Thanks for the awesome tool!

**Comment by borkdude** [comment]
@markus-wa In general, one should not rely on other namespaces to load certain libraries for you. E.g.: it's an implementation detail in clojure.core whether some of the already required namespaces loads `clojure.string`. You should explicitly require it, if you want to use that namespace.

**Comment by markus-wa** [comment]
Ah ok, that makes sense, thanks!
